libraries/state_history/trace_converter.cpp

```diff
@@ -30,25 +38,274 @@ void trace_converter::add_transaction(const transaction_trace_ptr& trace, const
+   fc::raw::pack(ds, obj);
+}
+
+template <typename Buffer, typename T, std::enable_if_t<has_context_free_segments<T>::value, int> = 0>
+void pack(Buffer& buffer, const T& obj) {
+   const auto comopressed_context_free_segments =
+       obj.context_free_segments.size() ? zlib_compress_bytes(fc::raw::pack(obj.context_free_segments)) : bytes{};
+   pack(buffer, std::make_pair(std::ref(obj.signatures), std::ref(comopressed_context_free_segments)));
+}
+
+void pack(bytes& buffer, const prunable_data_type& obj) {
+   buffer.push_back(obj.prunable_data.which());
+   obj.prunable_data.visit([&buffer](const auto& obj) { pack(buffer, obj); });
+}
+
+void pack(fc::datastream<char*>& ds, const prunable_data_type& obj) {
+   fc::raw::pack(ds, static_cast<uint8_t>(obj.prunable_data.which()));
+   obj.prunable_data.visit([&ds](const auto& obj) { pack(ds, obj); });
+}
+
+template <typename T, std::enable_if_t<!has_context_free_segments<T>::value, int> = 0>
+void unpack(const char*, fc::datastream<const char*>& ds, T& obj) {
+   fc::raw::unpack(ds, obj);
+}
+
+template <typename T, std::enable_if_t<has_context_free_segments<T>::value, int> = 0>
+void unpack(const char* read_buffer, fc::datastream<const char*>& ds, T& t) {
+   fc::raw::unpack(ds, t.signatures);
+   t.context_free_segments = unpack_zlib_compressed<decltype(t.context_free_segments)>(read_buffer, ds);
+}
 
-   return fc::raw::pack(make_history_context_wrapper(db, trace_debug_mode, traces));
+void unpack(const char* read_buffer, fc::datastream<const char*>& ds, prunable_data_type& prunable) {
+   uint8_t tag;
+   fc::raw::unpack(ds, tag);
+   prunable.prunable_data.set_which(tag);
+   prunable.prunable_data.visit([read_buffer, &ds](auto& v) { unpack(read_buffer, ds, v); });
+}
+
+/// used to traverse each trace along with its associated unpacked prunable_data
+template <typename Visitor>
+void visit_deserialized_trace(const char* read_buffer, fc::datastream<const char*>& ds, transaction_trace& trace,
+                              Visitor&& visitor) {
+   auto& trace_v0 = trace.get<transaction_trace_v0>();
+   if (trace_v0.failed_dtrx_trace.size()) {
+      // failed_dtrx_trace have at most one element because it is encoded as an optional
+      visit_deserialized_trace(read_buffer, ds, trace_v0.failed_dtrx_trace[0].recurse, std::forward<Visitor>(visitor));
+   }
+   if (trace_v0.partial) {
+      prunable_data_type prunable;
+      unpack(read_buffer, ds, prunable);
+      visitor(trace_v0, prunable);
+   }
+}
+
+struct restore_partial {
+   partial_transaction_v0& ptrx;
+   void                    operator()(prunable_data_type::full_legacy& data) const {
+      ptrx.signatures        = std::move(data.signatures);
+      ptrx.context_free_data = std::move(data.context_free_segments);
+   }
+   void operator()(prunable_data_type::none& data) const {}
+   void operator()(prunable_data_type::partial& data) const { ptrx.signatures = std::move(data.signatures); }
```

Comment:
We will need a new format for partial, for now I think this should EOS_ASSERT as not implemented.